graphistry/embed_utils.py

```diff
@@ -0,0 +1,579 @@
+import logging
+import numpy as np
+import pandas as pd
+from typing import Optional, Union, Callable, List, TYPE_CHECKING, Any, Tuple
+
+from .PlotterBase import Plottable
+from .compute.ComputeMixin import ComputeMixin
+
+def lazy_embed_import_dep():
+    try:
+        import torch
+        import torch.nn as nn
+        import dgl
+        from dgl.dataloading import GraphDataLoader
+        import torch.nn.functional as F
+        from .networks import HeteroEmbed
+        from tqdm import trange  # type: ignore
+        return True, torch, nn, dgl, GraphDataLoader, HeteroEmbed, F, trange
+
+    except:
+        return False, None, None, None, None, None, None, None
+
+
+if TYPE_CHECKING:
+    _, torch, _, _, _, _, _, _ = lazy_embed_import_dep()
+    TT = torch.Tensor
+    MIXIN_BASE = ComputeMixin
+else:
+    TT = Any
+    MIXIN_BASE = object
+    torch = Any
+
+XSymbolic = Optional[Union[List[str], str, pd.DataFrame]]
+ProtoSymbolic = Optional[Union[str, Callable[[TT, TT, TT], TT]]]  # type: ignore
+
+logging.StreamHandler.terminator = ""
+logger = logging.getLogger(__name__)
+#logger.setLevel(logging.DEBUG)
+
+
+def log(msg:str) -> None:
+    # setting every logs to WARNING level
+    logger.log(msg=msg, level=30)
+
+
+class EmbedDistScore:
+    @staticmethod
+    def TransE(h:TT, r:TT, t:TT) -> TT:  # type: ignore
+        return (h + r - t).norm(p=1, dim=1)  # type: ignore
+
+    @staticmethod
+    def DistMult(h:TT, r:TT, t:TT) -> TT:  # type: ignore
+        return (h * r * t).sum(dim=1)  # type: ignore
+
+    @staticmethod
+    def RotatE(h:TT, r:TT, t:TT) -> TT:  # type: ignore
+        return -(h * r - t).norm(p=1, dim=1)  # type: ignore
+
+
+class HeterographEmbedModuleMixin(MIXIN_BASE):
+    def __init__(self):
+        super().__init__()
+
+        self._protocol = {
```

Comment:
`._protocol` doesn't need to be in the instance, we can make a top-level `protocols`